presto-main/src/main/java/io/prestosql/sql/planner/ScopeAware.java

```diff
@@ -0,0 +1,156 @@
+        checkArgument(scope.isLocalScope(other.scope) || other.scope.isLocalScope(scope), "Expressions must be in the same local scope");
+
+        return equal(node, other.node);
+    }
+
+    @Override
+    public String toString()
+    {
+        return "ScopeAware(" + node + ")";
+    }
+
+    private boolean equal(Node left, Node right)
+    {
+        if (left instanceof Expression && right instanceof Expression) {
+            Expression leftExpression = (Expression) left;
+            Expression rightExpression = (Expression) right;
+            if (analysis.isColumnReference(leftExpression) && analysis.isColumnReference(rightExpression)) {
+                ResolvedField leftField = analysis.getResolvedField(leftExpression);
+                ResolvedField rightField = analysis.getResolvedField(rightExpression);
+
+                Scope leftScope = leftField.getScope();
+                Scope rightScope = rightField.getScope();
+
+                if (isLocalScope(leftScope) || isLocalScope(rightScope)) {
+                    // At least one of the fields comes from the same scope as the original expression, so compare by field id
+                    return leftField.getFieldId().equals(rightField.getFieldId());
+                }
+
+                // Otherwise, they are either:
+                // 1. Local to a subquery. In this case, the specific fields they reference don't matter as long as everything else is syntactically equal.
+                // 2. From a scope that is a parent to the base scope. In this case, the fields must necessarily be the same if they are syntactically equal.
+                return leftExpression.equals(rightExpression);
+            }
+        }
+
+        if (!left.shallowEquals(right)) {
+            return false;
+        }
+
+        List<? extends Node> leftChildren = left.getChildren();
+        List<? extends Node> rightChildren = right.getChildren();
+
+        if (leftChildren.size() != rightChildren.size()) {
+            return false;
+        }
+
+        for (int i = 0; i < leftChildren.size(); i++) {
+            if (!equal(leftChildren.get(i), rightChildren.get(i))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private int hash(Node node)
+    {
+        if (node instanceof Expression) {
+            Expression expression = (Expression) node;
+            if (analysis.isColumnReference(expression)) {
+                ResolvedField field = analysis.getResolvedField(expression);
+
+                Scope resolvedScope = field.getScope();
+                if (isLocalScope(resolvedScope)) {
+                    return field.getFieldId().hashCode();
+                }
+
+                return expression.hashCode();
+            }
+        }
+
+        List<? extends Node> children = node.getChildren();
+
+        if (children.isEmpty()) {
+            return node.hashCode();
+        }
+
+        int result = 1;
```

Comment:
this could be:
```
int result = node.hashCode()
```
to reduce collisions

Reply:
No, because `node.hashCode` is implemented recursively and we specifically don't want to consider children in the computation except via the call to `hash(...)` below. Note that the leaves will use `node.hashCode()`, so collisions for subtrees are going to be rare.

Reply:
nit: You could use `node.getClass().hashCode()` instead